src/Symfony/Component/Messenger/Transport/Sender/SendersLocator.php

```diff
@@ -56,4 +55,27 @@ public function getSenders(Envelope $envelope, ?bool &$handle = false): iterable
 
         $handle = $handle || null === $sender;
     }
+
+    /**
+     * @internal
+     */
+    protected function listTypes(Envelope $envelope): array
+    {
+        $class = \get_class($envelope->getMessage());
+
+        $parts = explode('\\', $class);
+        $tmp = array_shift($parts);
+        $classes = [];
+        foreach ($parts as $part) {
```

Comment:
the implemenation currently looks for more generic matches first e.g. `Acme\*` before `Acme\Foo\*`. It should be the other way round. First match on more concrete matches and then less ones to be consistent with the matching on full class name and then parents/interfaces.